File: rust/day2/src/main.rs
use std::fs;

fn main() {
    

    let mut final_score: i32 = 0;
    let mut final_score_predict: i32 = 0;

    for line in fs::read_to_string("data.txt").unwrap().lines() {
        let result = line.to_string();
        let mut value = result.chars();
        let pair: (char, char) = (value.next().unwrap(), value.last().unwrap());
        final_score += calc_score(pair);
        final_score_predict += calc_score_on_predict(pair.clone());
    }

    println!("final_score: {}", final_score);
    println!("final_score_predict: {}", final_score_predict);

}

fn calc_score(pair: (char, char)) -> i32 {
    let mut score_per_round: i32 = 0;

    /*
     * Rock X: 1
     * Paper Y: 2
     * Scissor Z: 3
     */

    match pair.0 {
        'A' => match pair.1 { // Rock
            'X' => score_per_round = 3 + 1, // draw
            'Y' => score_per_round = 6 + 2, // win
            'Z' => score_per_round = 0 + 3, // lose
            _ => eprintln!("error"),
        },
        'B' => match pair.1 { // Paper
            'X' => score_per_round = 0 + 1, // lose 
            'Y' => score_per_round = 3 + 2, // draw
            'Z' => score_per_round = 6 + 3, // win
            _ => eprintln!("error"),
        },
        'C' => match pair.1 { // Scissor
            'X' => score_per_round = 6 + 1, // win
            'Y' => score_per_round = 0 + 2, // lose
            'Z' => score_per_round = 3 + 3, // draw 
            _ => eprintln!("error"),
        }
        _ => eprintln!("error"),
    };

    score_per_round
}

fn calc_score_on_predict(pair: (char, char)) -> i32 {
    let mut score_per_round: i32 = 0;
    
    /*
     * Rock X: 1
     * Paper Y: 2
     * Scissor Z: 3
     */

    match pair.0 {
        'A' => match pair.1 { // Rock
            'X' => score_per_round = 0 + 3, // lose
            'Y' => score_per_round = 3 + 1, // draw
            'Z' => score_per_round = 6 + 2, // win
            _ => eprintln!("error"),
        },
        'B' => match pair.1 { // Paper
            'X' => score_per_round = 0 + 1, // lose
            'Y' => score_per_round = 3 + 2, // draw
            'Z' => score_per_round = 6 + 3, // win
            _ => eprintln!("error"),
        },
        'C' => match pair.1 { // Scissor
            'X' => score_per_round = 0 + 2, // lose 
            'Y' => score_per_round = 3 + 3, // draw
            'Z' => score_per_round = 6 + 1, // win 
            _ => eprintln!("error"),
        }
        _ => eprintln!("error"),
    };

    score_per_round
}

File: go/day2/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

type game struct {
	Rock     byte
	Paper    byte
	Scissors byte
}

type predict struct {
	Loss byte
	Draw byte
	Win  byte
}

const (
	ROCK     = 1
	PAPER    = 2
	SCISSORS = 3
	WON      = 6
	DRAW     = 3
	LOSS     = 0
)

func CalcScore(data []string) uint32 {

	enemy := game{
		Rock:     'A',
		Paper:    'B',
		Scissors: 'C',
	}

	player := game{
		Rock:     'X',
		Paper:    'Y',
		Scissors: 'Z',
	}
	var score uint32 = 0

	var scoreTable = map[byte]map[byte]uint32{
		enemy.Rock: {
			player.Paper:    WON + PAPER,
			player.Rock:     DRAW + ROCK,
			player.Scissors: LOSS + SCISSORS,
		},
		enemy.Scissors: {
			player.Paper:    LOSS + PAPER,
			player.Rock:     WON + ROCK,
			player.Scissors: DRAW + SCISSORS,
		},
		enemy.Paper: {
			player.Paper:    DRAW + PAPER,
			player.Rock:     LOSS + ROCK,
			player.Scissors: WON + SCISSORS,
		},
	}

	for _, line := range data {
		first := line[0]
		last := line[2]

		scoreMap, ok := scoreTable[first]
		if ok {
			value, ok := scoreMap[last]
			if ok {
				score += value
			} else {
				log.Fatalf("last not found in scoreMap: %c\n", last)
			}
		} else {
			log.Fatalf("first not found in scoreTable: %c\n", first)
		}
	}

	return score
}

func CalcScorePredict(data []string) uint32 {
	var score uint32 = 0
	enemy := game{
		Rock:     'A',
		Paper:    'B',
		Scissors: 'C',
	}

	decider := predict{
		Loss: 'X',
		Draw: 'Y',
		Win:  'Z',
	}

	var scoreTable = map[byte]map[byte]uint32{
		enemy.Rock: {
			decider.Loss: LOSS + SCISSORS,
			decider.Draw: DRAW + ROCK,
			decider.Win:  WON + PAPER,
		},
		enemy.Scissors: {
			decider.Loss: LOSS + PAPER,
			decider.Draw: DRAW + SCISSORS,
			decider.Win:  WON + ROCK,
		},
		enemy.Paper: {
			decider.Loss: LOSS + ROCK,
			decider.Draw: DRAW + PAPER,
			decider.Win:  WON + SCISSORS,
		},
	}

	for _, line := range data {
		first := line[0]
		last := line[2]

		scoreMap, ok := scoreTable[first]
		if ok {
			value, ok := scoreMap[last]
			if ok {
				score += value
			} else {
				log.Fatalf("last not found in scoreMap: %c\n", last)
			}
		} else {
			log.Fatalf("first not found in scoreTable: %c\n", first)
		}
	}

	return score
}

func main() {

	var score uint32 = 0
	file, err := os.Open("data.txt")

	if err != nil {
		log.Fatal(err)
	}
	defer func(file *os.File) {
		err := file.Close()
		if err != nil {
			log.Fatal("failed to close File")
		}
	}(file)

	scanner := bufio.NewScanner(file)
	var data []string

	for scanner.Scan() {
		data = append(data, scanner.Text())
	}

	score = CalcScorePredict(data)
	fmt.Printf("Predict Final Score: %v\n", score)

	score = CalcScore(data)
	fmt.Printf("Final Score: %v\n", score)

}
